app/src/main/java/org/mozilla/fenix/components/history/PagedHistoryProvider.kt

```diff
@@ -37,6 +37,11 @@ class DefaultPagedHistoryProvider(
 
     private var historyGroups: List<History.Group>? = null
 
+    /**
+     * Map of [History.Metadata] to their corresponding [History.Regular].
+     */
+    private var metadataToVisitMap = mutableMapOf<History.Metadata, History.Regular>()
```

Comment:
There is a concurrency issue here. The `getHistory` call on the provider is invoked on a background thread, and if multiple threads call it concurrently they may read the wrong data (from a wrong provider or wrong `getHistory` call). We'd have to use a different key (incl. offset/numberOfItems) and use a concurrent map, if we're keeping this approach.

Reply:
See my comment below though about mapping. I think we should not use a map here, but delete the closest visit by time.